### ISSUE TITLE
cnn using variable length data

### ISSUE THREAD

**Issue by manwinder123** [fchollet/keras#424]
So i have read issue #40 and i understand that theano's matrix's have to have the same dimensions.

I have Keras padding my sequences and that works. My data is just list of texts documents, its usually between 100 to 50k characters per document. I end up padding some documents with 49900 zeros (if text document is 100 characters long). I have a classification problem btw.

I switched to a Convolutional neural network because using a regular neural network with a few dense layers got way to slow when i increased the input sizes above 10k, also i thought the cnn would be able to notice that some documents have a ton of zeros, ignore that data and focus on the actual non-zero characters. Is my theory wrong?

Then the max pool would reduce the # of zeros even more.

I guess this is a sort of a feature request, i'm not sure how Keras deals with padding when its doing it thing but would it be possible for Keras to recognize the zeros and skip them?

Should i maybe using a different kind of neural network?

Thanks for any help :smile:

**Comment by pranv** [comment]
Have you tried using recurrent nets?

**Comment by manwinder123** [comment]
nope not yet, i'll look into it

**Comment by manwinder123** [comment]
So i tried all of the recurrent networks that Keras has to offer and they are way too slow. 
I take my inputs and reshape them using numpy (number of text documents, size of largest doc) <-- Shape

then i got my embedding layer with mask_zero set to True

got my rnn layer with return_sequences set to False (I have tried setting it to True)

finally my softmax activation layer.

one epoch ends up being around 30k seconds or something very large. I have tried all the rnn's avaliable, i messed around with the settings to see if the time per epoch would go down but i had no luck. I have switched to an autoencoder, i know that its not variable length but i thought it might give me a higher accuracy than the fully connected nn. 

So it seems that the autoencoder is pretty fast but its a little confusing. So i have set the data input size to 1000, anything after is ignored, anything will less than 1000 characters is filled in with zeros. 
Encoder: Dense(1000,500) -> Dense(500,250)
Decoder Dense(250,500) -> Dense(500,1000)

Softmax(250, 40) This is for when output_reconstruction is False
Softmax(1000, 40) this is for when output_reconstruction is True
I have 40 classifications btw

So from a quick look at autoencoders, i understand that they can be used for things like translation, english to french, that makes sense to me, input and output size should be similar. But i don't want to take a document and output it back out as some weird text. Should i maybe add some activation layers that take the output and shrink it down to 40, maybe add a convolution that has an output of 40 and i use the auto encoder on that?

**Comment by wjbianjason** [comment]
have you solved this question? I also need to use cnn dealing with varialble length input,do you hava any method that is no need to pad zeros or your theory that padding zeros is no influence for the result.I don't want to change the different kind of neural network btw.

**Comment by bizso09** [comment]
The underlying Theano implementation can handle variable length input without problems so this is a limitation of Keras itself.
